### PR TITLE
Fix #555 Python: Separate Python2 only articles into new Python2 Fathead

### LINKED ISSUE #555: Python: Separate Python2 only articles into new Python2 Fathead
### Description
Currently we have a single Python Fathead responsible for creating all articles related to Python 2 and Python 3. A subset of these articles are only relevant to Python2 and so we prepend the article title with "python2". This requires the user to enter a query of the form: `python2 {{query}}`.

This drastically reduces the flexibility of triggering Python2 articles because different people specify "python2" in different way. E.g. "python 2", "python 2.7", "python27", "python 27", etc.

Internally, Fathead have trigger words (just like Spice and Goodies), that can appear at the start or end of a query. This allows us to broaden the query space of a Fathead without modifying the article titles, or requiring redirects.

If we separate the `python2` articles into their own Fathead, we can specify as many "python2" trigger words as we want, and even require their presence in the query to trigger the Fathead. This will make it considerably easier to trigger the Python2 articles, and should have a big impact on Python coverage.

There are several ways to achieve a separate Python2 Fathead. The simplest being to copy the code from Python and adjust it to only handle Python2 articles.

However, we could definitely find a more clever approach: For example, the Python2 `fetch.sh` could simply execute the Python Fatheads `fetch.py` and `parse.py` scripts to get a copy of `output_py2.txt` and then rename that file to `output.txt`.

As long as DDG staff can `cd` into the Fathead's directory, execute `fetch.sh` and `parse.sh`, and end up with an `ouput.txt` in the current directory, we will be OK.

Does anyone have any other ideas or thoughts about how to achieve this?

Again, the end goal is to have a Fathead that covers articles that are specific to Python2 (i.e. no overlap with the articles in the Python Fathead), so that we can use Fathead trigger words to ensure a "python2" specifier is in the query. We can't achieve this with a unified Fathead because there is no other way of separating python2 articles from python3 articles when they have identical article titles.

### Example

| Query | Triggered Fathead | Trigger | "More at" Link | InfoBox Content | Currently Possible
--------|-------------------|---------|----------------|-----------------|-------------------
"[os.dup](https://duckduckgo.com/?q=os.dup)" | **Python** | None |points to python3 docs | link to "python2 os.dup" | Yes
"[python2 os.dup](https://duckduckgo.com/?q=python2+os.dup)" | **Python2** | "python2" | point to python2 docs | link to "python3 os.dup" | Yes
"[os.dup python 2.7](https://duckduckgo.com/?q=os.dup+python+2.7)" | **Python2** | "python 2.7" | points to python2 docs | link to "python3 os.dup" | NO
"[os.dup for python27](https://duckduckgo.com/?q=os.dup+for+python+27)" | **Python2** | "for python27" | points to python2 docs | link to "python3 os.dup" | NO

## People to notify
@pjhampton @duckduckgo/community-leaders @duckduckgo/language-leaders 

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a Pull Request

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

---

Instant Answer Page: https://duck.co/ia/view/python

### PR DESCRIPTION
**Pull request by Hestben**
## Description of new Instant Answer, or changes
First naive implementation suggestion.
Trigger fetch and parse in  python fathead, and retrieve the result for python2.
Caveat: the permutations of different order of words are not redirecting to the article, like the python3 articles. This is done in redirect, which combines python2 and python3 articles, and thus does not create a separate python2 output.

## Related Issues and Discussions

Fixes #555.

## People to notify

Maintainer:@gaulrobe
Creator of issue:@moollaza

---

Instant Answer Page: https://duck.co/ia/view/python2

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
@Hestben can you please make a new IA Page for ID `python2` -- This is a separate Fathead, so it requires its own IA Page before this can be deployed 👍

**Comment by Hestben** [comment]
@moollaza I created https://duck.co/ia/view/python2. I haven't worked with triggering before, so I haven't setup any triggers. Is the failed check something to worry about (https://travis-ci.org/duckduckgo/zeroclickinfo-fathead/builds/179809356)?

**Comment by Hestben** [comment]
I just found an error in the output.txt. There are three entries with the key "argparse ArgumentParser.add_argument_group". Thus, I think I need to run the redirect.py on the python2 output. I am working on that (an carefully checking the output I get).

**Comment by moollaza** [comment]
> Is the failed check something to worry about (https://travis-ci.org/duckduckgo/zeroclickinfo-fathead/builds/179809356)?

@Hestben nope, I've disabled travis. It should only run if/once we have a .travis.yml file in the repo 👍

**Comment by moollaza** [comment]
> There are three entries with the key "argparse ArgumentParser.add_argument_group".

@Hestben after using vim to sort unique entries in output.txt it dropped from ~16K entries to ~9.5K aricles, so it looks like we have a big problem with duplicates.

When I sorted it alphabetically it as immediately obvious we have several repeated entries, some even repeated more than once!

**Comment by moollaza** [comment]
@Hestben after cleaning up the output.txt I've deployed this to moollaza and beta.

On moollaza I've also managed to update the trigger words to work for the new python2 fathead.

now `python abs` triggers python, and `python 2.7 abs` triggers the python2 Fathead. That 2nd query can't show a result on DuckDuckGo.com right now, so this is great! We've added support for about 9000 python 2.x queries!

- https://moollaza.duckduckgo.com/?q=python2.7+abs&t=h_&ia=about
- https://moollaza.duckduckgo.com/?q=abs+python2.7&t=h_&ia=about

This is a good example of a query we've corrected:

https://duckduckgo.com/?q=_py_c_prod+python2&t=h_&ia=about
https://moollaza.duckduckgo.com/?q=_py_c_prod+python2&t=h_&ia=about

**Comment by moollaza** [comment]
@Hestben this is also on Beta, but the Python2 trigger words don't yet work, so triggering is a little difficult for now.

If you push the `output.txt` file to this PR it should automatically deploy the updated articles to Beta, and as a bonus you'll have a special trigger word for your Fathead, being your username:

So you'll be able to use https://beta.duckduckgo.com/?q=_py_c_prod+hestben 😄

**Comment by Hestben** [comment]
> @Hestben after using vim to sort unique entries in output.txt it dropped from ~16K entries to ~9.5K aricles, so it looks like we have a big problem with duplicates.

Yes, I also noticed that. And there are no permutations of the redirects, but I think I can manage to get redirect.py to run on the python2 results, which I believe can solve both of these problems.
I may have an improvement for redirect.py to get more permutations of the redirect to work, but I need to figure out if it is possible to rebase the pull request, otherwise, I will have to create a new PR for those improvements.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?FHBESNn8KC)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?q12hguVP0U)

**Comment by Hestben** [comment]
One strange thing. I get this error now:
`UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 206: ordinal not in range(128)`
I tried to use python3 instead, and that worked. What do you run when generating the output.txt? Could I just change the shell script to use python3?

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?uXLimWzu2p)

**Comment by Hestben** [comment]
I just pushed a change to redirect.py.
Now redirects with the same keys, with the same target will be added once, and not removed if that key happens to occur again in the input. And redirects with the same keys, creates a disambiguation, if the targets differ.
This will fix e.g.
https://duckduckgo.com/?q=python+dumps&ia=about
There are actually 5 different packages with the method 'dump'
https://docs.python.org/3.6/search.html?q=dumps
That the json.dump is displayed currently, is because there are an odd number of redirects for the key 'dumps' and the json one is the last to be considered. Before, on the first occurrence of a key, it would be added. If another entry wanted to add the same redirect key, the old key was removed. So, for every even number, the redirect key would be removed, and for every odd number, the redirect key would be added, and would point to the entry last considered.

This substantially increases the number of entries in output.txt:
`wc -l output.txt: 30239 output.txt`

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?HXNOO15unE)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?Whu8OAdEzO)

**Comment by moollaza** [comment]
> I tried to use python3 instead, and that worked. What do you run when generating the output.txt? Could I just change the shell script to use python3?

@Hestben you're welcome to use Python3

> Now redirects with the same keys, with the same target will be added once, and not removed if that key happens to occur again in the input. And redirects with the same keys, creates a disambiguation, if the targets differ.

>Now permutations like https://beta.duckduckgo.com/?q=python2+argparse+ArgumentParser+add_argument_group&ia=iatesting
and https://beta.duckduckgo.com/?q=python2+ArgumentParser+argparse+add_argument_group&ia=iatesting&iai=0
should work💃

This is awesome! Nicely done 👍  💯 

>This substantially increases the number of entries in output.txt:
wc -l output.txt: 30239 output.txt

That's no problem 👍  You can commit a removal of the output.txt once we're ready to merge the PR so it won't get added to the repo.

@Hestben can you please add a `trigger_words.txt` file as well? This will force your IA to have the associated trigger words:

**trigger_words.txt**:
```
pyhton
python
python2
python2.6
python26
python2.7
python27
python 27
python 2.4
python 2.5
python 2.6
python 2.7
python 2.7.10
python 2.7.12
python 2
python 2
```

**Comment by moollaza** [comment]
@Hestben it looks like the article for `fcntl` is being generated without a source_url value (the last tab) and It's causing install issues. Can you investigate why that one doesn't have a URL?

If there are any others, we should prevent them from being added to the output.txt unless we have a URL. 

FWIW, The `python` Fathead does produce that article with the correct link: https://duckduckgo.com/35/python2_fcntl?iax=1&ia=about

**Comment by Hestben** [comment]
> @Hestben can you please add a trigger_words.txt file as well? This will force your IA to have the associated trigger words:

I added the file trigger_words.txt. Will not the "python" and "pyhton" entries conflict with the original python fathead?

**Comment by Hestben** [comment]
> @Hestben it looks like the article for fcntl is being generated without a source_url value (the last tab) and It's causing install issues. Can you investigate why that one doesn't have a URL?

I have looked in the output, and have seen a bug in my code. Will fix that now.

**Comment by moollaza** [comment]
> I have looked in the output, and have seen a bug in my code. Will fix that now.

Awesome! Thanks 👍

**Comment by moollaza** [comment]
> I added the file trigger_words.txt. Will not the "python" and "pyhton" entries conflict with the original python fathead?

@Hestben nope, we're allowed multiple triggers for multiple Fatheads. Because the Python Fathead is older, it get's a first chance to trigger though.

So we'l see the `python` trigger word, and check if the remainder exists in the Python Fathead's articles, if not, we'll fallback to Python2 and so on

**Comment by Hestben** [comment]
> @Hestben nope, we're allowed overlapping triggers for multiple Fatheads. Seeing as the Python
Fathead is older (lower source_id), it get's the first chance to trigger though.

I see, good to know that the older fatheads get priority.

Unfortunately, after implementing my simple fix for the problem with the URL for fcntl, I have discovered a greater problem, that I am addressing now. It may take some days more before I check in a fix.

**Comment by Hestben** [comment]
I just checked in a major refactoring of redirect.py.
The python2 output.txt now contains
Articles:5339
Redirects:23962
Disambiguations:938

**Comment by Hestben** [comment]
I just commited a fix for the problem with disambiguations.
Now queries like [python2 cosh](https://beta.duckduckgo.com/?q=python2+cosh&ia=meanings&iai=%2F221%2Fcmath.cosh) work :dancer:

**Comment by Hestben** [comment]
Before this PR is merged, I would like to rebase it to squash some commits. We need to rebase it anyway, to remove the output.txt file anyway, right? So just tell me when it is good to go, and I will perform a rebase.

**Comment by moollaza** [comment]
> Before this PR is merged, I would like to rebase it to squash some commits.

@Hestben there's no need -- when I merge the PR it will be squashed automatically by GitHub 👍 

You can just commit the removal of the output.txt when this is ready for merge

**Comment by moollaza** [comment]
@Hestben this is looking awesome! I did a little testing and everything seems to be working correctly :)

However, one funny thing I noticed was "python2 print" -- we're showdig content from some Python Tutorials section, instead of the page related to Python built-in functions.

As a result we send them to a page that doesn't immediately mention the print function, and the Fathead doesn't provide a code block with example usage:

https://beta.duckduckgo.com/?q=python2+print&ia=about

The `python` Fathead seems to be using this page: https://docs.python.org/2.7/library/functions.html?highlight=built%20functions and sends the user there, which seems better

**Comment by moollaza** [comment]
If you'd like to fix that in another PR, I'm happy to merge this once you remove the `output.txt`.

Excellent work here, thanks a lot for helping us out and improving these Python searches 👍 💯 😃

File: lib/fathead/python2/parse.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+export PYTHONIOENCODING=utf8
+(cd ../python; ./parse.sh)
+#Run redirect on python2 results to get permutations of redirects, and remove
+# duplicates
+cp ../python/output_py2.txt ../python/output.txt
+(cd ../python; python redirect.py)
+cp ../python/output2.txt output.txt
+echo "Python 2 output contains"
+cat output.txt | ../python/output_statistics.awk
```
Comment (moollaza) [inline lib/fathead/python2/parse.sh]: Can you please add `LC_ALL=C sort output.txt -o output.txt` to sort the output file?

Comment (moollaza) [inline lib/fathead/python2/parse.sh]: or rather adjust `/python/parse.sh` to do that instead

**Comment by Hestben** [comment]
> However, one funny thing I noticed was "python2 print" -- we're showdig content from some Python Tutorials section, instead of the page related to Python built-in functions.

You are right. Unfortunately, the data we get for python2 from parse.py has this information, so to fix this, I need to do something to parse.py. I should probably do this in another PR, but it will be strange if the example of the IA is faulty. I'll look into it, to see if I can fix it in 5 minutes, else, I will open another PR for this.

File: lib/fathead/python2/parse.sh (hunk 1 above)
Comment (Hestben) [inline lib/fathead/python2/parse.sh]: Will do.

**Comment by Hestben** [comment]
> I'll look into it, to see if I can fix it in 5 minutes, else, I will open another PR for this.

Looking closer on the python 2.7 and python 3.5 docs, they differ in their tags, so I think I need to spend some time really researching how to improve the parsing so that the correct `print` will be parsed for python2. I was actually waiting for @jdorweiler to create an issue for other parse issues I found working on #538, I guess I will simply create an issue for that, and create a new PR. Maybe we should change the example query on the IA page to something that actually looks good?

**Comment by Hestben** [comment]
> If you'd like to fix that in another PR, I'm happy to merge this once you remove the output.txt.

@moollaza I removed output.txt in my latest commit. You can merge the PR now.

**Comment by moollaza** [comment]
> Maybe we should change the example query on the IA page to something that actually looks good?

Definitely. Any suggestions/preferences?

>@moollaza I removed output.txt in my latest commit. You can merge the PR now.

Excellent, thanks a lot 👍

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?EEJZWngpPA)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?J2HxhU2Q8d)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?i4burghkMo)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?o5wdoZmRu0)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=python2 print)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/python2_mobile.png?AcH6nuZnMh)

**Comment by Hestben** [comment]
Maybe we could use [python2 abs](https://duckduckgo.com/?q=python2+abs&t=ffsb&ia=about)
[Currently looks like this on beta](https://duckduckgo.com/?q=python2+abs&t=ffsb&ia=abou)

When installing this on the live, we could generate a new output.txt for the ordinary python fathead as well. That will add a lot of redirects and disambiguations.

**Comment by tagawa** [comment]
Hi @Hestben. Just letting you know that your PR has now been released: https://duckduckgo.com/?q=python2+abs&ia=about

Thanks again for another helpful contribution!

**Comment by Hestben** [comment]
@tagawa Hmm, why doesn't the queries 
https://duckduckgo.com/?q=python2+argparse+add_argument&t=ffsb&ia=iatesting
https://duckduckgo.com/?q=python2+add_argument&t=ffsb&ia=web
 work?
In the output.txt I get when I run the fathead, there are redirects for those queries.

**Comment by jdorweiler** [comment]
I suspect it's the underscores in the title. I haven't been able to find any redirects with underscores that work.

**Comment by Hestben** [comment]
@jdorweiler Aha. That seem like a major limitation for programming searches. Is that something internal in the duckduckgo-handling of the IA that doesn't work?

**Comment by jdorweiler** [comment]
@Hestben yeah it's internal. I made a task to look into it.

**Pull request merged**